src/test/lib/ApiTest.cpp

```diff
@@ -2126,3 +2126,195 @@ QuicTestCredentialLoad(const QUIC_CREDENTIAL_CONFIG* Config)
 
     TEST_QUIC_SUCCEEDED(Configuration.LoadCredential(Config));
 }
+
+void
+QuicTestStorage()
+{
+    const uint32_t SpecialInitialRtt = 55;
+
+#ifdef _KERNEL_MODE
+    DECLARE_CONST_UNICODE_STRING(GlobalStoragePath, L"\\Registry\\Machine\\System\\CurrentControlSet\\Services\\MsQuic\\Parameters\\");
+    DECLARE_CONST_UNICODE_STRING(AppStoragePath, L"\\Registry\\Machine\\System\\CurrentControlSet\\Services\\MsQuic\\Parameters\\Apps\\StorageTest\\");
+    DECLARE_CONST_UNICODE_STRING(ValueName, L"InitialRttMs");
+    HANDLE GlobalKey, AppKey;
+    OBJECT_ATTRIBUTES GlobalAttributes, AppAttributes;
+    InitializeObjectAttributes(
+        &GlobalAttributes,
+        (PUNICODE_STRING)&GlobalStoragePath,
+        OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+        NULL,
+        NULL);
+    InitializeObjectAttributes(
+        &AppAttributes,
+        (PUNICODE_STRING)&AppStoragePath,
+        OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+        NULL,
+        NULL);
+    TEST_QUIC_SUCCEEDED(
+        ZwOpenKey(
+            &GlobalKey,
+            KEY_READ | KEY_NOTIFY,
+            &GlobalAttributes));
+    ZwDeleteValueKey(
+        GlobalKey,
+        (PUNICODE_STRING)&ValueName);
+    if (QUIC_SUCCEEDED(
+        ZwOpenKey(
+            &AppKey,
+            KEY_READ | KEY_NOTIFY,
+            &AppAttributes))) {
+        ZwDeleteKey(AppKey);
+        ZwClose(AppKey);
+    }
+    TEST_QUIC_SUCCEEDED(
+        ZwCreateKey(
+            &AppKey,
+            KEY_READ | KEY_NOTIFY,
+            &AppAttributes,
+            0,
+            NULL,
+            REG_OPTION_NON_VOLATILE,
+            NULL));
+#elif _WIN32
+    RegDeleteKeyValueA(
+        HKEY_LOCAL_MACHINE,
+        "System\\CurrentControlSet\\Services\\MsQuic\\Parameters",
+        "InitialRttMs");
+    RegDeleteKeyA(
+        HKEY_LOCAL_MACHINE,
+        "System\\CurrentControlSet\\Services\\MsQuic\\Parameters\\Apps\\StorageTest");
+    HKEY Key;
+    DWORD Result =
+        RegCreateKeyA(
+            HKEY_LOCAL_MACHINE,
+            "System\\CurrentControlSet\\Services\\MsQuic\\Parameters\\Apps\\StorageTest",
+            &Key);
+    if (Result != NO_ERROR) {
+        //
+        // Skip test if not running under admin
+        //
+        return;
+    }
+    RegCloseKey(Key);
+#else
+    TEST_FAILURE("Storage tests not supported on this platform");
+#endif
+
+    MsQuicSettings Settings;
+
+    //
+    // Global settings (future TODO, its too late at this point to create the Parameters key)
```

Comment:
Let's check to see if the global key already existed, and only run these tests if it did. It will exist on WS2022+ builds.

Reply:
Oh, another option is that test.ps1 or run-gtest.ps1 can create the registry key before the test too. I think it'd be good to do both.

Reply:
I also need to bypass this test when registry write permission doesn't exist. But I'll work on that monday.

Reply:
Makes sense, but we currently run the tests as Admin in CI, so it's only a local problem.